### ISSUE TITLE
Order of flyd.on

### ISSUE THREAD

**Issue by irisjay** [paldepind/flyd#138]
```
var x = flyd.stream()
flyd.on(console.log.bind(console,'one'),x)
flyd.on(console.log.bind(console,'two'),x)
flyd.on(console.log.bind(console,'three'),x)
x('fire')
```

This gives the result
```
three fire
two fire
one fire
```

This seems counter intuitive to me and caused some issues for me. Is there any rationale for this behavior? Could we change it?

**Comment by irisjay** [comment]
Changing this section in `updateDeps(s)`
```
for (; orderNextIdx >= 0; --orderNextIdx) {
    o = order[orderNextIdx ];
    if (o.shouldUpdate === true) updateStream(o);
    o.queued = false;
  }
```
to this
```
  for (var x = 0; x <= orderNextIdx; x++) {
    o = order[x];
    if (o.shouldUpdate === true) updateStream(o);
    o.queued = false;
  }
```
seems to make it work as expected for me, though I'm not sure if there are any side consequences.

**Comment by dmitriz** [comment]
According to https://github.com/paldepind/flyd#flydonfn-s, the `on` method is essentially a "worse" version of `map` as it does the same but does not return anything useful. 

And if you use `map` instead of `on`, that behaviour becomes intuitive. Mapping over a function `f` means `f` is called with every stream value. So when you update your stream, all 3 functions are called. Does it make sense?

**Comment by StreetStrider** [comment]
@dmitriz, he notes on the order of called functions, not the function itself.

**Comment by dmitriz** [comment]
@StreetStrider Hm ... aren't they meant to fire simultaneously?

**Comment by StreetStrider** [comment]
@dmitriz what do you mean by «simultaneous» here?

**Comment by dmitriz** [comment]
@StreetStrider All at the same time, as soon as their parent emits.

**Comment by StreetStrider** [comment]
@dmitriz, yes, but this is irrelevant to what TS mentions. He just notes on that `on`-subscribed functions are called in reverse.

**Comment by dmitriz** [comment]
@StreetStrider The order is not guaranteed by the spec. It may depend on the implementation and on the runtime environment. Why does it matter?

**Comment by irisjay** [comment]
For side effects: clearly it could have great impact

**Comment by garkin** [comment]
This is a result of a topological sort of the dependency graph. It's needed for atomic updates.
<img src="https://user-images.githubusercontent.com/422774/32245548-a4d15344-beae-11e7-9523-a65817b17b41.png" width="150">
For example in this case sorted nodes array will look like `[OnBbb, bbb, OnCc, cc, bb, b, c, d]`.
Array with sort results should always be iterated in reverse to the build order due the nature of such sort. 
This is why you could not change order in the unwrapping loop. 

But you could change order in which we iterate through the children to build it (both `[listeners]` loop in the `updateDeps()` and in the `findDeps()`.
In this case you will get following order:
<img src="https://user-images.githubusercontent.com/422774/32246233-e68c6484-beb0-11e7-8583-e54c0e942318.png" width="150">
And sorted nodes: `[d, OnCc, cc, c, OnBb, bbb, bb, b]`. 
This will result in the same unwrapping order as a children order.

**Comment by StreetStrider** [comment]
Wow, nice picture.

**Comment by garkin** [comment]
The problem is a backward compatibilty.
It will change the order of execution for already existing programs. 
It will be breaking change.

**Comment by nordfjord** [comment]
This is a quirk of the library as is. Looking at the commit history the reverse iteration is because of a rather nice performance boost.

If you want some `b` to happen *after* some `a` you need to make it explicit by making `b` depend on `a`
e.g. for your original example.

```
const log = a => x => (console.log(a, x), x);
const x = stream();
x.map(log('first')).map(log('second')).map(log('third'));
x('fire');
```

Now in stead of the dependency graph looking like this:

```
   x
 / | \
1  2  3
```

It looks like
```
x
|
1
|
2
|
3
```

Meaning things will happen in your preferred order.

I'm not sure this is something we'd like to change, given it's here for perf reasons.
